File: src/tests/dap/fixtures/test_program.c
/*
 * Test program for DAP integration tests — C variant.
 *
 * Exercises DAP variable inspection across every DAP type:
 *   - Scalars:   int, long, float, double, char, _Bool
 *   - Strings:   const char*, char[]
 *   - Arrays:    int[], double[]
 *   - Structs:   simple (Counter), with nested members (AdapterInfo)
 *   - Nested:    struct-of-structs (Connection containing Counter + AdapterInfo)
 *   - Pointers:  int*, struct*, nested pointers
 *   - Enums:     error_kind + error_code
 *   - Unions:    variant payload (value_or_message)
 *   - Volatile:  volatile int (tests that debugger handles volatile correctly)
 *   - Typedefs:  counter_t alias
 *
 * Intended to be run with lldb-dap or gdb.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

/* ── scalars, strings, arrays ─────────────────────────────────────── */

static int number = 42;
static long big_number = 2147483648L;
static float ratio = 3.14f;
static double pi = 3.141592653589793;
static char letter = 'X';
static int flag = 1;                    /* _Bool / int */
static const char *greeting = "Hello, DAP!";
static char buffer[64] = "initialized buffer with trailing garbage";

static int items[] = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 12, 15, 20};
static size_t item_count = sizeof(items) / sizeof(items[0]);

static double coords[] = {1.1, 2.2, 3.3, 4.4, 5.5};
static size_t coord_count = sizeof(coords) / sizeof(coords[0]);

/* ── enum + union ──────────────────────────────────────────────────── */

typedef enum {
    ERR_NONE = 0,
    ERR_TIMEOUT,
    ERR_DISCONNECTED,
    ERR_INVALID,
} error_kind;

typedef union {
    long   numeric;
    char   message[64];
} payload;

/* ── structs + typedef ─────────────────────────────────────────────── */

typedef struct {
    int    value;
    double threshold;
    const char *label;
} Counter;

typedef struct {
    const char   *name;
    int           major;
    int           minor;
    int           patch;
    int           connected;           /* bool-like */
} AdapterInfo;

typedef struct {
    Counter         counter;
    AdapterInfo     adapter;
    int             port;
    error_kind      last_error;
    payload         last_payload;
} Connection;

/* ── recursive function (deep stack trace) ──────────────────────────── */

long factorial(long n) {
    if (n <= 1) return 1;
    return n * factorial(n - 1);
}

/* ── loop with conditional (conditional bp target) ─────────────────── */

void process_items(const int *src, size_t count, int *dst) {
    for (size_t i = 0; i < count; i++) {
        int doubled = src[i] * 2;       /* conditional bp: src[i] > 10 */
        dst[i] = doubled;
    }
}

/* ── nested calls (step-in / step-out target) ─────────────────────── */

int inner(int x) {
    int square = x * x;
    return square;
}

int middle(int x) {
    int y = x + 3;
    int z = inner(y);
    return z + 1;
}

int outer(void) {
    int result = middle(5);
    return result * 2;
}

/* ── volatile test ──────────────────────────────────────────────────── */

static volatile int heartbeat = 0;

/* ── main ──────────────────────────────────────────────────────────── */

int main(void) {
    int doubled[20] = {0};
    volatile int local_volatile = 99;

    /* ── complex struct assembly ──────────────────────────────────── */
    Counter c = {.value = 10, .threshold = 0.5, .label = "main-counter"};

    AdapterInfo info = {
        .name      = "debugpy",
        .major     = 1,
        .minor     = 8,
        .patch     = 13,
        .connected = 1,
    };

    Connection conn = {
        .counter     = c,
        .adapter     = info,
        .port        = 5678,
        .last_error  = ERR_TIMEOUT,
        .last_payload.message = "connection attempt timed out after 30 seconds",
    };

    heartbeat = 1;  /* [bp-1a] after volatile write — inspect heartbeat */

    int *heap_number = malloc(sizeof(int));
    if (heap_number) *heap_number = 999;
    int *heap_array = malloc(sizeof(int) * 4);
    if (heap_array) {
        heap_array[0] = 10;
        heap_array[1] = 20;
        heap_array[2] = 30;
        heap_array[3] = 40;
    }

    /* [bp-1] inspect: number, greeting, items[3], c.value, c.label,
     *   info.name, info.major, conn.port, conn.adapter.name,
     *   conn.last_error, conn.last_payload.message,
     *   *heap_number, heap_array[2], heartbeat, local_volatile */

    printf("number = %d\n", number);
    printf("pi    = %.15f\n", pi);
    printf("text  = %s\n", greeting);
    printf("flag  = %d\n", flag);
    printf("info  = %s v%d.%d.%d\n", info.name, info.major, info.minor, info.patch);
    printf("conn.port = %d\n", conn.port);

    /* loop: step_over friendly */
    process_items(items, item_count, doubled);
    printf("doubled[0] = %d, doubled[last] = %d\n",
           doubled[0], doubled[item_count - 1]);

    /* [bp-2] after loop: inspect doubled[3], coords[2], strerror, etc. */

    /* recursion */
    long fact = factorial(5);
    printf("factorial(5) = %ld\n", fact);

    /* object mutation */
    c.value += 1;
    c.value += 1;
    printf("counter.value = %d\n", c.value);

    /* [bp-3] after counter ops: inspect c.value,  c.threshold */

    /* nested calls */
    int outer_result = outer();
    printf("outer_result = %d\n", outer_result);

    /* [bp-4] near end: inspect outer_result, heap_number */

    int x = 10;
    int y = 20;
    int z = x + y;
    printf("z = %d\n", z);

    free(heap_number);
    free(heap_array);
    return 0;
}

File: src/tests/dap/fixtures/test_program.js
/**
 * Test program for DAP integration tests — JavaScript variant.
 *
 * Exercises: launch with stopOnEntry, line breakpoints, continue,
 * step over/into, stack trace, variable inspection, expression
 * evaluation.  Intended to be run with the bundled dap_node_adapter.js
 * or any JavaScript DAP adapter.
 */

// ── Counter class for object inspection ──────────────────────────
class Counter {
    constructor(start = 0) {
        this.value = start;
        this.label = "counter";
    }

    increment() {
        this.value++;
        return this.value;
    }
}

// ── recursive function for deeper stack traces ────────────────────
function factorial(n) {
    if (n <= 1) return 1;
    return n * factorial(n - 1);
}

// ── loop with conditional — exercise conditional breakpoints ──────
function processItems(items) {
    const results = [];
    for (const item of items) {
        const doubled = item * 2;  // conditional bp: item > 10
        results.push(doubled);
    }
    return results;
}

// ── nested calls for step_in / step_out ───────────────────────────
function inner(x) {
    const square = x * x;
    return square;
}

function middle(x) {
    const y = x + 3;
    const z = inner(y);
    return z + 1;
}

function outer() {
    const result = middle(5);
    return result * 2;
}

// ── main ───────────────────────────────────────────────────────────
function main() {
    // basic types to inspect
    const text    = "Hello, DAP!";
    const number  = 42;
    const pi      = 3.14159;
    const flag    = true;

    const items   = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 12, 15, 20];
    const mapping = { key_a: 100, key_b: 200 };
    const counter = new Counter(10);

    // [bp-1] inspect locals
    console.log(`text   = ${text}`);
    console.log(`number = ${number}`);
    console.log(`pi     = ${pi}`);
    console.log(`flag   = ${flag}`);

    // loop
    const doubled = processItems(items);
    console.log(`doubled[0] = ${doubled[0]}, size = ${doubled.length}`);

    // [bp-2] after loop

    // recursion
    const fact = factorial(5);
    console.log(`factorial(5) = ${fact}`);

    // object mutation
    counter.increment();
    counter.increment();
    console.log(`counter.value = ${counter.value}`);

    // [bp-3] after counter ops

    // nested calls
    const outerResult = outer();
    console.log(`outer = ${outerResult}`);

    // [bp-4] near end

    const x = 10;
    const y = 20;
    const z = x + y;
    console.log(`z = ${z}`);
}

main();
